### ISSUE TITLE
UTF8Encoding reports incorrect error position during incremental decoding

### ISSUE THREAD

**Issue by BCSharp** [dotnet/corefx#37969]
TL;DR
In .NET Core 3.0.100-preview5-011568, during UTF-8 incremental decoding, when an invalid byte is located in one increment but can only be detected during decoding of the subsequent increment, the reported index position of the invalid byte refers to a different byte.

Full story: consider the following input to be decoded:

```csharp
var bytes = new byte[] { (byte)'x', 0xED, (byte)'y', (byte)'z'};
```

It contains one invalid byte in position 2 (zero-based-index value of 1) surrouned by ASCII-range bytes. When this input is decoded in a single pass by the `UTF8Encoding`, the encoding provides the correct value to its decoder fallback:

```csharp
var encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);

try
{
    encoding.GetChars(bytes);
}
catch (DecoderFallbackException ex)
{
    Console.WriteLine($"ex.Index: {ex.Index} (should be 1)");
    Console.WriteLine($"first byte in ex.BytesUnknown: {ex.BytesUnknown[0]:X2} (should be ED)");
}

```
Output:
```
ex.Index: 1 (should be 1)
first byte in ex.BytesUnknown: ED (should be ED)
```

Now the same input is being decoded in two increments, each two bytes long:

```csharp
var chars = new char[4];
var decoder = encoding.GetDecoder();
int count;

// Increment 1: bytes { 'x', 0xED }
count = decoder.GetCharCount(bytes, index: 0, count: 2, flush: false);
Console.WriteLine($"char count: {count}");
count = decoder.GetChars(bytes, byteIndex: 0, byteCount: 2, chars, charIndex: 0, flush: false);
Console.WriteLine($"num decoded chars: {count}");
```

The output is:

```
char count: 1
num decoded chars: 1
```

which is correct: although the invalid byte is within the first increment, no fallback can be triggered yet because it may have been a start of a valid UTF-8 sequence (byte ED can be a start of an encoded surrogate point). Therefore, only character 'x' is decoded in this step.

The issue is with Increment 2:

```csharp
// Increment 2: bytes { 'y', 'z' }
try
{
    count = decoder.GetCharCount(bytes, index: 2, count: 2, flush: true);
}
catch (DecoderFallbackException ex)
{
    Console.WriteLine($"ex.Index: {ex.Index} (should be -1)");
    Console.WriteLine($"first byte in ex.BytesUnknown: {ex.BytesUnknown[0]:X2} (should be ED)");
}
```

The output is:

```
ex.Index: 0 (should be -1)
first byte in ex.BytesUnknown: ED (should be ED)
```
`BytesUnknown` correctly carries the value of the invalid byte form the previous increment, however the index value, which is an offset relative to the beginning of the decoding buffer, incorrectly points to byte 'y'. A correct value of -1 indicates that the error position lies within the previous increment, one byte back from the end (i.e. the last byte of the previous increment). This was the behaviour of .NET Core 3.0.0-preview4-27615-11. Also all other .NET frameworks report a negative index in such case.

Consistent index reporting is essential for statefull decoder fallback implementations to be able to distinguish cases like `{ 'x', 0xED } + {'y', 'z' }` and `{'x', 'y'} + { 0xED, 'z'}`

cc @GrabYourPitchforks

**Comment by GrabYourPitchforks** [comment]
After a brief investigation I think this change can be constrained to just the EncoderNLS and DecoderNLS classes. My gut says this should be a very small change, but I'll need to set aside some time to work on the tests. Interesting find!

**Comment by joperezr** [comment]
@GrabYourPitchforks this is marked as 3.0, do you still plan on working on this today?

**Comment by GrabYourPitchforks** [comment]
While testing this I came across an interesting behavior of Full Framework that seems like a bug.

```cs
var decoder = new UTF8Encoding(
    encoderShouldEmitUTF8Identifier: false,
    throwOnInvalidBytes: true).GetDecoder();

// Start of four-byte sequence.

var charsWritten = decoder.GetChars(new byte[] { 0xF1, 0x80 }, 0, 2, new char[10], 0, false);
Debug.Assert(charsWritten == 0);

try
{
    // Terminate four-byte sequence two bytes in, start new ASCII sequence.

    charsWritten = decoder.GetCharCount(new byte[] { 0x60 }, 0, 1, false);
}
catch (DecoderFallbackException ex)
{
    Debug.Assert(ex.Index == -2); // expected = -2, actual = -4
}
```

I'll investigate a bit more to see if this is the intended behavior or if it's truly a bug in Full Framework. If it's a bug then we won't try to preserve the buggy behavior for Core.

**Comment by GrabYourPitchforks** [comment]
@joperezr - This probably won't be in for noon tomorrow, but the PR should be ready this week (probably tomorrow afternoon). I'll email shiproom when the time comes, as this bug meets the post-6/26 criteria.

**Comment by BCSharp** [comment]
> I'll investigate a bit more to see if this is the intended behavior or if it's truly a bug in Full Framework. If it's a bug then we won't try to preserve the buggy behavior for Core.

I believe this is truly a bug in Full Framework that has already been reported (#29898) and fixed in Core 3.0.0-preview4-27615-11. 

Note that the incorrect index is only happening for `GetCharCount` (or `GetChar` overloads that depend on `GetCharCount`), which underreports the index by the length of `BytesUnknown`. I agree with @slozier that it is possibly due to backtracking the input buffer pointer twice (maybe first in `GetChar` and then again in `GetCharCount`).
